glib-macros/src/clone.rs

```diff
@@ -1,952 +1,487 @@
+                            }
+                        };
+                    }
+                }
+            },
+            CaptureKind::WeakAllowNone => quote! {
+                let #alias = #crate_ident::clone::Upgrade::upgrade(&#alias);
+            },
+            _ => Default::default(),
         }
-        parts.next();
     }
-    let ret = match ret.as_str() {
-        "strong" => BorrowKind::Strong,
-        "weak" => BorrowKind::Weak,
-        "weak-allow-none" => BorrowKind::WeakAllowNone,
-        "to-owned" => BorrowKind::ToOwned,
-        "default-return" => {
-            return Err(parts
-                .generate_error_with_span("`@default-return` should be after `=>`", start_span));
-        }
-        "default-panic" => {
-            return Err(
-                parts.generate_error_with_span("`@default-panic` should be after `=>`", start_span)
-            );
-        }
-        k => {
-            return Err(parts.generate_error_with_span(
-                &format!(
-                    "Unknown keyword `{k}`, only `weak`, `weak-allow-none`, `to-owned` and \
-                    `strong` are allowed"
-                ),
-                start_span,
-            ));
-        }
-    };
-    Ok(ret)
 }
 
-fn parse_ident(
-    parts: &mut PeekableProcIter,
-    elements: &mut Vec<ElemToClone>,
-) -> Result<(), TokenStream> {
-    let borrow_kind = keyword(parts)?;
-    let name = full_ident(parts, borrow_kind)?;
-    let name_span = parts.current_span;
-    if name.ends_with('.') {
-        return Err(
-            parts.generate_error_with_span(&format!("Invalid variable name: `{name}`"), name_span)
-        );
-    }
-    let alias = match parts.peek() {
-        Some(TokenTree::Ident(p)) if p.to_string() == "as" => {
-            parts.next();
-            let current_span = parts.current_span;
-            match parts.next() {
-                Some(TokenTree::Ident(i)) => Some(i.to_string()),
-                Some(x) => {
-                    let err = format!("Expected ident after `as` keyword, found `{x}`");
-                    return Err(parts.generate_error(&err));
+#[derive(Clone)]
+enum ClosureOrAsync {
+    Closure(ExprClosure),
+    Async(ExprAsync),
+}
+
+impl ToTokens for ClosureOrAsync {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        match self {
+            ClosureOrAsync::Closure(ref c) => c.to_tokens(tokens),
+            ClosureOrAsync::Async(ref a) => a.to_tokens(tokens),
+        }
+    }
+}
+
+struct Clone {
+    captures: Vec<Capture>,
+    default_return: DefaultReturn,
+    body: ClosureOrAsync,
+}
+
+impl Parse for Clone {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let mut captures: Vec<Capture> = vec![];
+        let mut default_return: DefaultReturn = DefaultReturn::default();
+
+        loop {
+            // Parse none, one or more attributes.
+            //
+            // There must either be one or no attributes here. Multiple attributes are not
```

Comment:
It contradicts your sentence just above which says "none, one or more attributes".